source/idf/EthernetDevice.cpp

```diff
@@ -0,0 +1,162 @@
+            }
+            break;
+        }
+
+        std::cout << "[IDF::EthernetDevice] Connected to " << serverName << ":" << serverPort << std::endl;
+
+        Manageable::open();
+    }
+}
+
+/** disconnects from the server */
+void EthernetDevice::close() {
+    if (mOpen) {
+        ::close(socketHandle);
+        Manageable::close();
+    }
+}
+
+std::vector<std::vector<unsigned char> > EthernetDevice::read() {
+    std::vector<std::vector<unsigned char> > results;
+    std::vector<unsigned char> buffer(packetLength);
+
+    while (read(&buffer[0], buffer.size())) {
+        results.push_back(buffer);
+    }
+
+    return results;
+}
+
+int EthernetDevice::read(unsigned char *buffer, size_t length) {
+    if (!mOpen) {
+        open();
+    }
+
+    int bytesRecvd = 0;
+    unsigned bytesTotal = 0;
+    unsigned char readBuff[length];
+    memset(&readBuff, 0, length);
+
+    while(bytesTotal < length) {
+        // non-blocking receive
+        bytesRecvd = recvfrom(socketHandle, &readBuff[bytesTotal], length-bytesTotal, MSG_DONTWAIT, (struct sockaddr*)&srcAddr, &srcAddrLen);
+        if (bytesRecvd < 0) {
+            if (errno == EAGAIN) { return 0; } // no data, try again later
+            else if (errno == EINTR) { continue; } // interrupted, retry
+            else {
+                close();
+                throw IOException("Error while reading " + name + ": " + strerror(errno));
+            }
+        } else if (bytesRecvd > 0) {
+            bytesTotal += static_cast<unsigned>(bytesRecvd);
+        }
+    }
+    memcpy(buffer, &readBuff, length);
+    return bytesTotal;
+}
+
+int EthernetDevice::peek(unsigned char *buffer, size_t length) {
+    if (!mOpen) {
+        open();
+    }
+
+    int bytesRecvd = 0;
+    unsigned bytesTotal = 0;
+    unsigned char readBuff[length];
+    memset(&readBuff, 0, length);
+
+    while(bytesTotal < length) {
+        // non-blocking receive
+        bytesRecvd = recvfrom(socketHandle, &readBuff[bytesTotal], length-bytesTotal, MSG_DONTWAIT | MSG_PEEK, (struct sockaddr*)&srcAddr, &srcAddrLen);
+        if (bytesRecvd < 0) {
+            if (errno == EAGAIN) { return 0; } // no data, try again later
+            else if (errno == EINTR) { continue; } // interrupted, retry
+            else {
+                close();
+                throw IOException("Error while reading " + name + ": " + strerror(errno));
+            }
+        } else if (bytesRecvd > 0) {
+            bytesTotal += static_cast<unsigned>(bytesRecvd);
+        }
+    }
+    memcpy(buffer, &readBuff, length);
+    return bytesTotal;
+}
+
+int EthernetDevice::write(const void *buffer, size_t length) {
+    if (!mOpen) {
+        open();
+    }
+
+    int bytesSent = 0;
+    unsigned bytesTotal = 0;
+
+    while (bytesTotal < length) {
+        bytesSent = sendto(socketHandle, (&buffer)[bytesTotal], length-bytesTotal, MSG_NOSIGNAL, (struct sockaddr *)&serverAddr, serverAddrLen);
```

Comment:
for tcp, what is expected for the user do for a partial sendto? for udp, we can check that the length and bytesSent should be equal.

Reply:
good question. That would end up being dependent on specific end devices, right? I'll reach out to our device vendor. We can use their expectation as the default. Since its virtual it can be overriden for subsequent devices.

Reply:
so.. i tend to treat sends different than recv's in terms of retries and looping. I assume the packet must be sent immediately. one option would be to add a flag to pass into this write method that indicates whether to loop until all bytes are sent except on an actual error. one could do the same thing with recv but it is usually a little easier to handle partial recv's and wait until the next attempt than a partial send.